### PR TITLE
Fix partition

### PR DESCRIPTION
**Pull request by Rubtsowa**
test DPNP_QUEUE_GPU=1 pytest "tests/test_sort.py::test_partition[[3, 4, 2, 1]-float64-0]" was aborted on GPU. In this PR bag was fixed.

### CODE REVIEW

File: dpnp/backend/kernels/dpnp_krnl_sorting.cpp [hunk 1]
```diff
@@ -87,45 +87,70 @@ template <typename _DataType>
...
+        std::partial_sort(matrix, matrix + shape_[ndim-1], matrix + shape_[ndim-1]);
+        for (size_t j = ind_begin; j < ind_end + 1; ++j)
+        {
+            size_t ind = j - ind_begin;
+            arr[j] = matrix[ind];
+        }
+    }
+
+    size_t* shape = reinterpret_cast<size_t*>(dpnp_memory_alloc_c(ndim * sizeof(size_t)));
+    std::copy(shape_, shape_ + ndim, shape);
```
Comment (samir-nasibli) [inline dpnp/backend/kernels/dpnp_krnl_sorting.cpp:131]: try this:
```suggestion
    DPNP_QUEUE.memcpy(shape, shape_, ndim * sizeof(size_t));
```

File: dpnp/backend/kernels/dpnp_krnl_sorting.cpp [hunk 2]
```diff
@@ -87,45 +87,70 @@ template <typename _DataType>
...
+        }
+        std::partial_sort(matrix, matrix + shape_[ndim-1], matrix + shape_[ndim-1]);
+        for (size_t j = ind_begin; j < ind_end + 1; ++j)
+        {
+            size_t ind = j - ind_begin;
+            arr[j] = matrix[ind];
+        }
+    }
+
+    size_t* shape = reinterpret_cast<size_t*>(dpnp_memory_alloc_c(ndim * sizeof(size_t)));
```
Comment (samir-nasibli) [inline dpnp/backend/kernels/dpnp_krnl_sorting.cpp:130]: But actually you don't use this at all.

File: dpnp/backend/kernels/dpnp_krnl_sorting.cpp [hunk 3]
```diff
@@ -87,45 +87,70 @@ template <typename _DataType>
 class dpnp_partition_c_kernel;
 
 template <typename _DataType>
-void dpnp_partition_c(const void* sort_array1_in, void* result1, const size_t kth, const size_t* shape, const size_t ndim)
+void dpnp_partition_c(void* array1_in, void* result1, const size_t kth, const size_t* shape_, const size_t ndim)
 {
```
Comment (samir-nasibli) [inline dpnp/backend/kernels/dpnp_krnl_sorting.cpp:92]: Again and again. No nullptr check. Provide please.

**Comment by samir-nasibli** [comment]
@Rubtsowa 
Provide please short description about the fix. What was fixed? which tests were enabled or added?

File: dpnp/backend/kernels/dpnp_krnl_sorting.cpp [hunk 4]
```diff
@@ -87,33 +87,63 @@ template <typename _DataType>
...
+        for (size_t j = ind_begin; j < ind_end + 1; ++j)
+        {
+            size_t ind = j - ind_begin;
+            arr[j] = matrix[ind];
+        }
+    }
+
+    size_t* shape = reinterpret_cast<size_t*>(dpnp_memory_alloc_c(ndim * sizeof(size_t)));
+    DPNP_QUEUE.memcpy(shape, shape_, ndim * sizeof(size_t));
+    std::copy(shape_, shape_ + ndim, shape);
```
Comment (densmirn) [inline dpnp/backend/kernels/dpnp_krnl_sorting.cpp:137]: `std::copy` is not needed more.
```suggestion
```

File: dpnp/backend/kernels/dpnp_krnl_sorting.cpp (hunk 2 above)
Comment (samir-nasibli) [reply dpnp/backend/kernels/dpnp_krnl_sorting.cpp:135]: outdated.

**Review by densmirn — DISMISSED** [review]

File: dpnp/backend/kernels/dpnp_krnl_sorting.cpp [hunk 5]
```diff
@@ -87,33 +87,62 @@ template <typename _DataType>
...
+            size_t ind = j - ind_begin;
+            matrix[ind] = arr[j];
+        }
+        std::partial_sort(matrix, matrix + shape_[ndim-1], matrix + shape_[ndim-1]);
+        for (size_t j = ind_begin; j < ind_end + 1; ++j)
+        {
+            size_t ind = j - ind_begin;
+            arr[j] = matrix[ind];
+        }
+    }
```
Comment (samir-nasibli) [inline dpnp/backend/kernels/dpnp_krnl_sorting.cpp:133]: You can try parallelize this loop by adding `parallel_for` kernel in the future.

Also `std::partial_sort` is supported by `pstl` (see https://software.intel.com/content/www/us/en/develop/articles/get-started-with-parallel-stl.html) and `dpc++` + `pstl` as well (`dpl`, see: https://spec.oneapi.com/versions/0.5.0/oneAPI/Elements/onedpl/onedpl_root.html#).

This is just suggestions and it's better to do such optimization work on other PR(s).

**Review by samir-nasibli — APPROVED** [review]

**Review by densmirn — APPROVED** [review]

**Pull request merged**